### PR TITLE
[ogre1.x] Add support for wide angle camera

### PR DESCRIPTION
**Pull request by iche033**
# 🎉 New feature

## Summary

Ported the wide angle camera feature from gazebo-classic. Added the base classes and ogre 1.x implementation. 

## Test it

Added a wide_angle_camera integration test and a demo.

Run the integration test with `ogre` render engine from your colcon workspace:

```
RENDER_ENGINE_VALUES=ogre ./build/ignition-rendering7/bin/INTEGRATION_wide_angle_camera
```

Build and run the [wide_angle_camera](https://github.com/ignitionrobotics/ign-rendering/tree/wide_angle_camera_ogre/examples/wide_angle_camera) demo:

wide_angle_camera_demo

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

File: include/ignition/rendering/CameraLens.hh [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+      /// \param[in] _other The other camera lens
+      public: CameraLens &operator=(const CameraLens &_other);
+
+      /// \internal
+      /// \brief Converts projection type from one of presets to `custom`
+      private: void ConvertToCustom();
+
+      /// \internal
+      /// \brief Private data pointer
+      private: std::unique_ptr<CameraLensPrivate> dataPtr;
```
Comment (scpeters) [inline include/ignition/rendering/CameraLens.hh:185]: I think we could add `ignition-utils` as a dependency so we can use the `ImplPtr`? I'll open an issue to do this in parallel, without blocking

File: examples/wide_angle_camera/Main.cc [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    {
+      CameraPtr camera = createCamera(engineName);
+      if (camera)
+      {
+        cameras.push_back(camera);
+      }
+    }
+    catch (...)
+    {
+      // std::cout << ex.what() << std::endl;
```
Comment (ahcorde) [inline examples/wide_angle_camera/Main.cc:184]: remove ?

File: examples/wide_angle_camera/GlutWindow.cc [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+unsigned int g_cameraIndex = 0;
+ir::ImagePtr g_image;
+ignition::common::ConnectionPtr g_connection;
+
+bool g_initContext = false;
+
+#if __APPLE__
+  CGLContextObj g_context;
+  CGLContextObj g_glutContext;
+#elif _WIN32
```
Comment (ahcorde) [inline examples/wide_angle_camera/GlutWindow.cc:60]: Add windows support ? here a reference https://github.com/ignitionrobotics/ign-rendering/pull/291

File: include/ignition/rendering/CameraLens.hh [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+      /// \brief Sets whether the image should be scaled to fit horizontal FOV
+      /// If True, the projection will compute a new focal length for achieving
+      /// the desired FOV
+      /// \param[in] _scale true if it should,
+      ///   note: c1 and f parameters are ignored in this case
+      public: void SetScaleToHFOV(bool _scale);
+
+      /// \brief Apply mapping function to input number
+      /// \return Result of the application
```
Comment (ahcorde) [inline include/ignition/rendering/CameraLens.hh:167]:
```suggestion
      \param[in] _f
      /// \return Result of the application
```

File: include/ignition/rendering/WideAngleCamera.hh [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      /// \return Screen coordinates. Z is the distance of point from camera
+      /// optical center.
+      public: virtual math::Vector3d Project3d(const math::Vector3d &_pt) const
+           = 0;
+
+      /// \brief Subscribes a new listener to this camera's new frame event
+      /// \param[in] _listener New camera listener callback
+      public: virtual common::ConnectionPtr ConnectNewWideAngleFrame(
+          std::function<void(const unsigned char *, unsigned int, unsigned int,
+          unsigned int, const std::string &)>  _subscriber) = 0;
```
Comment (ahcorde) [inline include/ignition/rendering/WideAngleCamera.hh:62]:
```suggestion
          unsigned int, const std::string &)>  _listener) = 0;
```

File: ogre/include/ignition/rendering/ogre/OgreWideAngleCamera.hh [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+      /// \brief Gets the environment texture size
+      /// \return Texture size
+      public: unsigned int EnvTextureSize() const;
+
+      /// \brief Sets environment texture size
+      /// \param[in] _size Texture size
+      public: void SetEnvTextureSize(int _size);
+
+      /// \brief Set the background color of the camera
+      /// \param[in] _color Color to set the background to
```
Comment (ahcorde) [inline ogre/include/ignition/rendering/ogre/OgreWideAngleCamera.hh:78]:
```suggestion
      /// \param[in] _color Color to set the background to
      /// \return
```

File: ogre/include/ignition/rendering/ogre/OgreWideAngleCamera.hh [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+      // Documentation inherited.
+      public: virtual void PreRender() override;
+
+      /// \brief Implementation of the render call
+      public: virtual void Render() override;
+
+      // Documentation inherited
+      public: common::ConnectionPtr ConnectNewWideAngleFrame(
+          std::function<void(const unsigned char *, unsigned int, unsigned int,
+          unsigned int, const std::string &)>  _subscriber) override;
```
Comment (ahcorde) [inline ogre/include/ignition/rendering/ogre/OgreWideAngleCamera.hh:109]: is this listener o subscriber ?
```suggestion
          unsigned int, const std::string &)>  _subscriber) override;
```

File: ogre/src/OgreWideAngleCamera.cc [hunk 8]
```diff
@@ -0,0 +1,662 @@
...
+  public: Ogre::PixelFormat envCubeMapTextureFormat;
+
+  /// \brief A single cube map texture
+  public: Ogre::Texture *envCubeMapTexture = nullptr;
+
+  /// \brief Output texture
+  public: Ogre::Texture *ogreRenderTexture = nullptr;
+
+  /// \brief Pointer to material, used for second rendering pass
+  public: Ogre::MaterialPtr compMat;
```
Comment (ahcorde) [inline ogre/src/OgreWideAngleCamera.cc:71]:
```suggestion
  public: Ogre::MaterialPtr compMat{nullptr};
```

File: ogre/src/OgreWideAngleCamera.cc [hunk 9]
```diff
@@ -0,0 +1,662 @@
...
+};
+
+using namespace ignition;
+using namespace rendering;
+
+//////////////////////////////////////////////////
+OgreWideAngleCamera::OgreWideAngleCamera()
+    : dataPtr(std::make_unique<OgreWideAngleCameraPrivate>())
+{
+  for (int i = 0; i < 6; ++i)
```
Comment (ahcorde) [inline ogre/src/OgreWideAngleCamera.cc:101]: define `6` as a constant ? it's used multiples times in this file

File: ogre/src/OgreWideAngleCamera.cc [hunk 10]
```diff
@@ -0,0 +1,662 @@
...
+void OgreWideAngleCamera::SetEnvTextureSize(int _size)
+{
+  this->dataPtr->envTextureSize = _size;
+}
+
+/////////////////////////////////////////////////
+void OgreWideAngleCamera::CreateCamera()
+{
+  // Create dummy ogre camera object
+  Ogre::SceneManager *ogreSceneManager  = this->scene->OgreSceneManager();
```
Comment (ahcorde) [inline ogre/src/OgreWideAngleCamera.cc:200]:
```suggestion
  Ogre::SceneManager *ogreSceneManager = this->scene->OgreSceneManager();
```

File: ogre/src/OgreWideAngleCamera.cc [hunk 11]
```diff
@@ -0,0 +1,662 @@
...
+    this->ogreNode->attachObject(this->dataPtr->envCameras[i]);
+
+  // set environment cameras orientation
+  this->dataPtr->envCameras[0]->yaw(Ogre::Degree(-90));
+  this->dataPtr->envCameras[1]->yaw(Ogre::Degree(90));
+  this->dataPtr->envCameras[2]->pitch(Ogre::Degree(90));
+  this->dataPtr->envCameras[3]->pitch(Ogre::Degree(-90));
+  this->dataPtr->envCameras[5]->yaw(Ogre::Degree(180));
+}
+
```
Comment (ahcorde) [inline ogre/src/OgreWideAngleCamera.cc:258]:
```suggestion
```

File: ogre/src/OgreWideAngleCamera.cc [hunk 12]
```diff
@@ -0,0 +1,662 @@
...
+    this->dataPtr->envCameras[i]->yaw(Ogre::Degree(-90.0));
+    this->dataPtr->envCameras[i]->roll(Ogre::Degree(-90.0));
+
+    this->dataPtr->envCameras[i]->setNearClipDistance(nearPlane);
+    this->dataPtr->envCameras[i]->setFarClipDistance(farPlane);
+  }
+
+  for (int i = 0; i < 6; ++i)
+    this->ogreNode->attachObject(this->dataPtr->envCameras[i]);
```
Comment (ahcorde) [inline ogre/src/OgreWideAngleCamera.cc:249]:
```suggestion
  for (int i = 0; i < 6; ++i)
  {
    this->ogreNode->attachObject(this->dataPtr->envCameras[i]);
  }
```

File: src/CameraLens.cc [hunk 13]
```diff
@@ -0,0 +1,387 @@
...
+  public: double c2 = 1.0;
+
+  /// \brief Angle offset factor
+  public: double c3 = 0.0;
+
+  /// \brief Linear scale factor, may be adjusted in runtime
+  public: double f = 1.0;
+
+  /// \brief Visible field of view
+  public: double cutOffAngle = IGN_PI*0.5;
```
Comment (ahcorde) [inline src/CameraLens.cc:45]:
```suggestion
  public: double cutOffAngle = IGN_PI * 0.5;
```

File: test/integration/wide_angle_camera.cc [hunk 14]
```diff
@@ -0,0 +1,383 @@
...
+
+//////////////////////////////////////////////////
+/// \brief callback to get the wide angle camera image data
+void OnNewWideAngleFrame(const unsigned char *_data,
+                    unsigned int _width, unsigned int _height,
+                    unsigned int /*_channels*/,
+                    const std::string &/*_format*/)
+{
+  g_mutex.lock();
+  auto bufferSize = _width * _height * 3;
```
Comment (ahcorde) [inline test/integration/wide_angle_camera.cc:70]: ?
```suggestion
  auto bufferSize = _width * _height * _channels;
```

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: include/ignition/rendering/CameraLens.hh (hunk 1 above)
Comment (scpeters) [reply include/ignition/rendering/CameraLens.hh:185]: this dependency has been added in #492, so I recommend using `ImplPtr` in this new class

Comment (iche033) [reply include/ignition/rendering/CameraLens.hh]: switched to use `ImplPtr` in 21d9f79

File: examples/wide_angle_camera/Main.cc (hunk 2 above)
Comment (iche033) [reply examples/wide_angle_camera/Main.cc]: removed. 3399e6d

File: include/ignition/rendering/CameraLens.hh (hunk 4 above)
Comment (iche033) [reply include/ignition/rendering/CameraLens.hh:167]: 3399e6d

File: include/ignition/rendering/WideAngleCamera.hh (hunk 5 above)
Comment (iche033) [reply include/ignition/rendering/WideAngleCamera.hh:62]: changed to `_subscriber`. 3399e6d

File: ogre/include/ignition/rendering/ogre/OgreWideAngleCamera.hh (hunk 6 above)
Comment (iche033) [reply ogre/include/ignition/rendering/ogre/OgreWideAngleCamera.hh:77]: 3399e6d

File: ogre/src/OgreWideAngleCamera.cc (hunk 11 above)
Comment (iche033) [reply ogre/src/OgreWideAngleCamera.cc:262]: 3399e6d

File: ogre/src/OgreWideAngleCamera.cc (hunk 10 above)
Comment (iche033) [reply ogre/src/OgreWideAngleCamera.cc]: 3399e6d

File: ogre/src/OgreWideAngleCamera.cc (hunk 12 above)
Comment (iche033) [reply ogre/src/OgreWideAngleCamera.cc]: 3399e6d

File: test/integration/wide_angle_camera.cc (hunk 14 above)
Comment (iche033) [reply test/integration/wide_angle_camera.cc]: 3399e6d

File: src/CameraLens.cc (hunk 13 above)
Comment (iche033) [reply src/CameraLens.cc]: 3399e6d

File: ogre/include/ignition/rendering/ogre/OgreWideAngleCamera.hh (hunk 7 above)
Comment (iche033) [reply ogre/include/ignition/rendering/ogre/OgreWideAngleCamera.hh:111]: changed to `_subscriber`. 3399e6d

File: ogre/src/OgreWideAngleCamera.cc (hunk 8 above)
Comment (iche033) [reply ogre/src/OgreWideAngleCamera.cc:74]: this is an `Ogre::SharedPtr`, which unlike `std::shared_ptr, can not be set to `nullptr`. So I kept it the way it is.

File: ogre/src/OgreWideAngleCamera.cc (hunk 9 above)
Comment (iche033) [reply ogre/src/OgreWideAngleCamera.cc]: added constant. 3399e6d

File: examples/wide_angle_camera/GlutWindow.cc (hunk 3 above)
Comment (iche033) [reply examples/wide_angle_camera/GlutWindow.cc:67]: adapted windows changes to the CMakeList and cpp files in this example. a826e9e

**Review by WilliamLewww — APPROVED** [review]

File: src/CameraLens.cc [hunk 15]
```diff
@@ -0,0 +1,387 @@
...
+
+            /// \brief Get the angle transformation function
+            /// \return The same function which was passed to constructor
+            public: AngleFunctionType AngleFunction() const
+            {
+              return std::get<0>(value);
+            }
+
+            /// \brief Apply function to float value
+            /// \return The result of application
```
Comment (ahcorde) [inline src/CameraLens.cc:109]:
```suggestion
            /// param[in] _t
            /// \return The result of application
```

File: ogre/src/OgreWideAngleCamera.cc [hunk 16]
```diff
@@ -0,0 +1,666 @@
...
+
+      double f = this->Lens().F();
+      double fov = this->HFOV().Radian();
+      // recompute f if scale to HFOV is true
+      if (this->Lens().ScaleToHFOV())
+      {
+        double param = (fov/2.0) / this->Lens().C2() + this->Lens().C3();
+        double funRes = this->Lens().ApplyMappingFunction(
+            static_cast<float>(param));
+        f = 1.0/(this->Lens().C1()*funRes);
```
Comment (ahcorde) [inline ogre/src/OgreWideAngleCamera.cc:549]:
```suggestion
        f = 1.0 / (this->Lens().C1() * funRes);
```
```suggestion
        f = 1.0/(this->Lens().C1()*funRes);
```

File: test/integration/wide_angle_camera.cc [hunk 17]
```diff
@@ -0,0 +1,383 @@
...
+    ignerr << "Engine '" << _renderEngine
+           << "' doesn't support wide angle cameras" << std::endl;
+    return;
+  }
+
+  auto *engine = ignition::rendering::engine(_renderEngine);
+  if (!engine)
+  {
+    ignerr << "Engine '" << _renderEngine
+              << "' was unable to be retrieved" << std::endl;
```
Comment (ahcorde) [inline test/integration/wide_angle_camera.cc:97]:
```suggestion
    ignerr << "Engine '" << _renderEngine
           << "' was unable to be retrieved" << std::endl;
```

File: src/CameraLens.cc [hunk 18]
```diff
@@ -0,0 +1,387 @@
...
+    this->ConvertToCustom();
+
+  try
+  {
+    this->dataPtr->fun = CameraLens::Implementation::MapFunctionEnum(_fun);
+  }
+  catch(const std::exception &ex)
+  {
+    ignerr << "`Fun` value [" << _fun << "] is not known, "
+          << "keeping the old one" << std::endl;
```
Comment (ahcorde) [inline src/CameraLens.cc:354]:
```suggestion
    ignerr << "`Fun` value [" << _fun << "] is not known, "
           << "keeping the old one" << std::endl;
```

File: src/CameraLens.cc [hunk 19]
```diff
@@ -0,0 +1,387 @@
...
+
+    try
+    {
+      this->dataPtr->fun =
+          CameraLens::Implementation::MapFunctionEnum(std::get<4>(params));
+    }
+    catch(const std::exception &ex)
+    {
+      ignerr << "`fun` value [" << std::get<4>(params)
+            << "] is not known, keeping the old one" << std::endl;
```
Comment (ahcorde) [inline src/CameraLens.cc:300]:
```suggestion
      ignerr << "`fun` value [" << std::get<4>(params)
             << "] is not known, keeping the old one" << std::endl;
```

**Review by ahcorde — APPROVED** [review]
just some small style comments, otherwise LGTM

File: examples/wide_angle_camera/CMakeLists.txt [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+
+if (WIN32)
+  find_package(FreeGLUT REQUIRED)
+  set(TARGET_THIRD_PARTY_DEPENDS ${TARGET_THIRD_PARTY_DEPENDS} FreeGLUT::freeglut)
+endif()
+
+add_executable(wide_angle_camera Main.cc GlutWindow.cc)
+
+target_link_libraries(wide_angle_camera
+#  ${GLUT_LIBRARIES}
```
[the remark this answers is not in the document]
Comment (iche033) [reply examples/wide_angle_camera/CMakeLists.txt]: removed. e6d23741e29a535327f6dfec5b83894379c2a1c1

File: src/CameraLens.cc (hunk 15 above)
Comment (iche033) [reply src/CameraLens.cc:110]: e6d23741e29a535327f6dfec5b83894379c2a1c1

File: test/integration/wide_angle_camera.cc (hunk 17 above)
Comment (iche033) [reply test/integration/wide_angle_camera.cc]: e6d23741e29a535327f6dfec5b83894379c2a1c1

File: ogre/src/OgreWideAngleCamera.cc (hunk 16 above)
Comment (iche033) [reply ogre/src/OgreWideAngleCamera.cc]: e6d23741e29a535327f6dfec5b83894379c2a1c1

File: src/CameraLens.cc (hunk 19 above)
Comment (iche033) [reply src/CameraLens.cc]: e6d23741e29a535327f6dfec5b83894379c2a1c1

**Pull request merged**